code/app/src/main/java/com/example/bread/repository/ParticipantRepository.java

```diff
@@ -115,4 +163,73 @@ public void checkIfAlreadyFollowed(@NonNull String username, String followerUser
+        WriteBatch batch = db.batch();
+
+        DocumentReference senderRef = db.collection("participants").document(senderId);
+        DocumentReference receiverRef = db.collection("participants").document(receiverId);
+        DocumentReference requestRef = db.collection("followRequests").document(requestId);
+
+        // Delete the follow request
+        batch.delete(requestRef);
+
+        // Conditionally add to following and followers list if not already added
+        senderRef.get().addOnSuccessListener(senderSnapshot -> {
+            List<String> following = (List<String>) senderSnapshot.get("following");
+            if (following == null || !following.contains(receiverId)) {
+                batch.update(senderRef, "following", FieldValue.arrayUnion(receiverId));
+            }
+
+            receiverRef.get().addOnSuccessListener(receiverSnapshot -> {
+                List<String> followers = (List<String>) receiverSnapshot.get("followers");
+                if (followers == null || !followers.contains(senderId)) {
+                    batch.update(receiverRef, "followers", FieldValue.arrayUnion(senderId));
+
+                    // Commit the batch
+                    batch.commit().addOnSuccessListener(onSuccessListener).addOnFailureListener(onFailureListener);
+                } else {
+                    onFailureListener.onFailure(new Exception("Already following."));
+                }
+            }).addOnFailureListener(onFailureListener);
+        }).addOnFailureListener(onFailureListener);
+    }
+
+    public void declineFollowRequest(String requestId, OnSuccessListener<Void> onSuccessListener, OnFailureListener onFailureListener) {
+        DocumentReference requestRef = db.collection("followRequests").document(requestId);
+        requestRef.delete().addOnSuccessListener(onSuccessListener).addOnFailureListener(onFailureListener);
+    }
+
+    public void updateFollowCounts(String userId, boolean isIncrement, boolean isFollowers) {
+        DocumentReference userRef = db.collection("participants").document(userId);
+        db.runTransaction(transaction -> {
+                    DocumentSnapshot snapshot = transaction.get(userRef);
+                    long newCount = snapshot.getLong(isFollowers ? "followersCount" : "followingCount") + (isIncrement ? 1 : -1);
+                    transaction.update(userRef, isFollowers ? "followersCount" : "followingCount", newCount);
+                    return null;
+                }).addOnSuccessListener(aVoid -> Log.d("ParticipantRepository", "Count updated successfully."))
+                .addOnFailureListener(e -> Log.e("ParticipantRepository", "Error updating count: " + e.getMessage()));
+    }
+
+    public void unfollowUser(String currentUserId, String unfollowUserId, OnSuccessListener<Void> onSuccessListener, OnFailureListener onFailureListener) {
```

Comment:
With the current setup this method won't work. Above comments apply here as well.